File: Python solutions/14. Longest Common Prefix.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        # vertical scan 
        # time complexity: o(n) 
        # space complexity: o(1)
        # if strs == None or len(strs) == 0: return ""
        # for i in range(len(strs[0])):
        #     for j in range(len(strs)):
        #         if (i == len(strs[j]) or strs[j][i] != strs[0][i]): 
        #             return strs[0][:i]
        # return strs[0]
        
        # using zip in python 
        # Create one iterator per string using zip, it will stop at the shortest string
	# s is a tuple of characters at current position for each string
	# create a set to test unicity
        idx = 0
        for s in zip(*strs):
            if len(set(s)) != 1:
                break
            else:
                idx += 1
        return strs[0][:idx] if idx > 0 else ""

File: Java solutions/14. Longest Common Prefix.java
class Solution {
    public String longestCommonPrefix(String[] strs) {
        // one pass vertical scanning
        // time complexity: o(k + n) k is the lenfth of the first word, n is the length of strs 
        // space complexity: o(1) in place
        if (strs.length == 0 || strs == null) return ""; // edge case
        for (int i = 0; i < strs[0].length(); i++){
            // loop through each char in the first word
            for (int j = 1; j < strs.length; j++){
                // compare with every other str in strs
                if (i == strs[j].length() || strs[j].charAt(i) != strs[0].charAt(i)){
                    return strs[0].substring(0, i);
                }
            }
        }
        return strs[0];
    }
}
